### PR TITLE
Added EDMCOverlay support for notifications

### PR DESCRIPTION
**Pull request by DeadMor0z**
Hello!
I've added support for [EDMCOverlay](https://github.com/inorton/EDMCOverlay) EDMC plugin. If EDMCOverlay is not accessible, standart notification will be shown.
Looks great in game (look at the left side of screenshot).

Desktop Screenshot 2020 05 10 - 20 05 42 05

### CODE REVIEW

**Comment by Xjph** [comment]
That _does_ look great!

I've got a couple of other small changes and bugfixes I want to wrap up, than I'll get this merged and into an official build. Hopefully this coming weekend!

File: Observatory/EDOverlay.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        private string headerId = "obsHeader";
+        private string bodyId = "obsBody";
+
+        private bool ConnectOverlay()
+        {
+            bool result = true;
+            try
+            {
+                sock = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                try
```
Comment (Xjph) [inline Observatory/EDOverlay.cs:23]: Can the port be probed to see what is listening there instead? Relying on an exception is a bit gross.

File: Observatory/EDOverlay.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                {
+                    TcpListener listener = new TcpListener(new IPAddress(new byte [] {127, 0, 0, 1}), 5010);
+                    listener.Start();
+                    listener.Stop();
+                    result = false;
+                }
+                catch (ArgumentNullException ane)
+                {
+                    Console.WriteLine("ArgumentNullException : {0}", ane.ToString());
```
Comment (Xjph) [inline Observatory/EDOverlay.cs:32]: Only making this comment once, but there is a lot of console output. This isn't a console application. Exceptions should be either handled quietly or reported to the user through the GUI.

File: Observatory/EDOverlay.cs (hunk 1 above)
Comment (DeadMor0z) [reply Observatory/EDOverlay.cs:23]: There is no need to put socket creation in try/catch block - there could not be any exception.

I didn't find any good (and fast) way to see if port is open or not. We can try to directly connect to port to see if it is open, but it takes time. As far as I know - the only exception which can happen when trying to listen port is when port is used - and that is what we need.

Comment (Xjph) [reply Observatory/EDOverlay.cs:23]: Aside from the awkwardness of relying on listener.Start() to throw an exception if the port is used, there's also the edge case where something other than EDMCOverlay is listening on port 5010. I checked EDMCOverlay and it doesn't provide any identifying response at all. Might need to invoke netstat to check the process name that's attached to the port, which is, unfortunately, also kind of gross.

I'll accept as-is, but if I find any way to ensure that it's EDMCO on the other side that isn't a giant pain in the ass I'll make the changes myself before release.

File: Observatory/EDOverlay.cs (hunk 2 above)
Comment (DeadMor0z) [reply Observatory/EDOverlay.cs:32]: Well, it's up to you. Personally I like them to be, as they help to pin down unexpected situation. Users will not see console output anyway, unless they start program from command line.

**Comment by DeadMor0z** [comment]
I've made a "netstat"-like way of finding out EDMCOverlay listening port. Works good and fast enough.

**Comment by Xjph** [comment]
After having a day or so to think about it, I love this functionality.

I am not a huge fan of EDMCOverlay's particular implementation.

I'm going to finish up a few of my own changes, then decide if I want to use this as is, or just implement my own transparent overlay window not involving any external software.

**Comment by DeadMor0z** [comment]
I don't think having multiple overlays is a good idea.

**Comment by Xjph** [comment]
> I don't think having multiple overlays is a good idea.

What problem do you see with that exactly? Keep in mind the EDMCOverlay's "overlay" is simply a windows form with its OnPaint methods overridden, and differs from Observatory's notification popup on purely a cosmetic basis.

**Comment by Xjph** [comment]
Alright, going to accept this more or less as-is after thinking on it some more. I might replace the functionality in the future .NET Core rewrite, but it'll do for now.

**Pull request merged**